RuntimeWarning from "scalar divide" in AsymptoticCalculator.teststatistic when sqrtqmuA_v == 0

### Summary

Similar to Issue https://github.com/scikit-hep/pyhf/issues/1872, `RuntimeWarning` are being surfaced from exactly zero being computed due to operating system level changes(?).

<details>
<summary>Details:</summary>

```
______________________________ test_toms748_scan _______________________________

hypotest_args = (1.0, [51, 48, 277.77777777777777, 55.183673469387756], <pyhf.pdf.Model object at 0x11dbd2330>)

    @pytest.mark.usefixtures("tmp_path")
    def test_toms748_scan(hypotest_args):
        """
        Test the upper limit toms748 scan returns the correct structure and values
        """
        _, data, model = hypotest_args
>       results = pyhf.infer.intervals.upper_limits.toms748_scan(
            data, model, 0, 5, rtol=1e-8
        )

_          = 1.0
data       = [51, 48, 277.77777777777777, 55.183673469387756]
hypotest_args = (1.0, [51, 48, 277.77777777777777, 55.183673469387756], <pyhf.pdf.Model object at 0x11dbd2330>)
model      = <pyhf.pdf.Model object at 0x11dbd2330>

tests/test_infer.py:32: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/pyhf/infer/intervals/upper_limits.py:121: in toms748_scan
    lower_results = f_cached(bounds_low)
                    ^^^^^^^^^^^^^^^^^^^^
        _exp_idx   = 1
        atol       = 2e-12
        best_bracket = <function toms748_scan.<locals>.best_bracket at 0x11f45b740>
        bounds_low = 0
        bounds_up  = 5
        cache      = {}
        data       = [51, 48, 277.77777777777777, 55.183673469387756]
        f          = <function toms748_scan.<locals>.f at 0x11f4aa980>
        f_cached   = <function toms748_scan.<locals>.f_cached at 0x11f4ab560>
        from_upper_limit_fn = False
        hypotest_kwargs = {}
        level      = 0.05
        model      = <pyhf.pdf.Model object at 0x11dbd2330>
        rtol       = 1e-08
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/pyhf/infer/intervals/upper_limits.py:86: in f_cached
    cache[poi] = hypotest(
        cache      = {}
        data       = [51, 48, 277.77777777777777, 55.183673469387756]
        hypotest_kwargs = {}
        model      = <pyhf.pdf.Model object at 0x11dbd2330>
        poi        = 0
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/pyhf/infer/__init__.py:167: in hypotest
    teststat = calc.teststatistic(poi_test)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        calc       = <pyhf.infer.calculators.AsymptoticCalculator object at 0x11dbd76f0>
        calctype   = 'asymptotics'
        data       = [51, 48, 277.77777777777777, 55.183673469387756]
        fixed_params = [False, False, False]
        init_pars  = [1.0, 1.0, 1.0]
        kwargs     = {}
        par_bounds = [(0, 10), (1e-10, 10.0), (1e-10, 10.0)]
        pdf        = <pyhf.pdf.Model object at 0x11dbd2330>
        poi_test   = 0
        return_calculator = False
        return_expected = False
        return_expected_set = True
        return_tail_probs = False
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/pyhf/infer/calculators.py:444: in teststatistic
    teststat = tensorlib.conditional(
        _          = <pyhf.optimize.scipy_optimizer object at 0x11ee4a700>
        _false_case = <function AsymptoticCalculator.teststatistic.<locals>._false_case at 0x11f478400>
        _true_case = <function AsymptoticCalculator.teststatistic.<locals>._true_case at 0x11f4798a0>
        asimov_data = array([ 50.15240199,  50.05755656, 278.62445548,  53.12230492])
        asimov_mu  = 0.0
        asimov_mubhathat = array([0.        , 1.00304804, 0.96264532])
        mubhathat  = array([0.        , 1.00304804, 0.96264532])
        mubhathat_A = array([0.        , 1.00304529, 0.96260976])
        muhatbhat  = array([0.        , 1.00305119, 0.96266962])
        muhatbhat_A = array([1.08420217e-19, 1.00304879e+00, 9.62633637e-01])
        poi_test   = 0
        qmuA_v     = array(0.)
        qmu_v      = array(1.2923573e-07)
        self       = <pyhf.infer.calculators.AsymptoticCalculator object at 0x11dbd76f0>
        sqrtqmu_v  = np.float64(0.00035949371298171903)
        tensorlib  = <pyhf.tensor.numpy_backend.numpy_backend object at 0x11f4d5540>
        teststat_func = <function qmu_tilde at 0x10ec17b00>
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/pyhf/tensor/numpy_backend.py:199: in conditional
    return true_callable() if predicate else false_callable()
                                             ^^^^^^^^^^^^^^^^
        false_callable = <function AsymptoticCalculator.teststatistic.<locals>._false_case at 0x11f478400>
        predicate  = np.False_
        self       = <pyhf.tensor.numpy_backend.numpy_backend object at 0x11f4d5540>
        true_callable = <function AsymptoticCalculator.teststatistic.<locals>._true_case at 0x11f4798a0>
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def _false_case():
        qmu = tensorlib.power(sqrtqmu_v, 2)
        qmu_A = tensorlib.power(self.sqrtqmuA_v, 2)
>       return (qmu - qmu_A) / (2 * self.sqrtqmuA_v)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E       RuntimeWarning: divide by zero encountered in scalar divide

qmu        = np.float64(1.292357296733826e-07)
qmu_A      = np.float64(0.0)
self       = <pyhf.infer.calculators.AsymptoticCalculator object at 0x11dbd76f0>
sqrtqmu_v  = np.float64(0.00035949371298171903)
tensorlib  = <pyhf.tensor.numpy_backend.numpy_backend object at 0x11f4d5540>

/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/pyhf/infer/calculators.py:441: RuntimeWarning
------------------------------ Captured log setup ------------------------------
INFO     pyhf.pdf:pdf.py:788 Validating spec against schema: model.json
INFO     pyhf.pdf:pdf.py:469 adding modifier mu (1 new nuisance parameters)
INFO     pyhf.pdf:pdf.py:469 adding modifier uncorr_bkguncrt (2 new nuisance parameters)
```

</details>

### OS / Environment

```console
macos-latest
```

### Steps to Reproduce


<!--- ...or if you have a failing CLI command paste it between the quotes below -->
```console (paste below)
pytest tests/test_infer.py
```


### File Upload (optional)

_No response_

### Expected Results

No failure

### Actual Results

```console
>       return (qmu - qmu_A) / (2 * self.sqrtqmuA_v)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E       RuntimeWarning: divide by zero encountered in scalar divide
```

### pyhf Version

```console
dev
```

### Code of Conduct

- [x] I agree to follow the Code of Conduct

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -176,7 +176,7 @@ filterwarnings = [
     'ignore: Exception ignored in:pytest.PytestUnraisableExceptionWarning',  #FIXME: Exception ignored in: <_io.FileIO [closed]>
     'ignore:invalid value encountered in (true_)?divide:RuntimeWarning',  #FIXME
     'ignore:invalid value encountered in add:RuntimeWarning',  #FIXME
-    'ignore:divide by zero encountered in (scalar |true_)?divide:RuntimeWarning',  #FIXME: pytest tests/test_tensor.py::test_pdf_calculations[numpy]. Issue #2722
+    'ignore:divide by zero encountered in (true_)?divide:RuntimeWarning',  #FIXME: pytest tests/test_tensor.py::test_pdf_calculations[numpy]
     'ignore:[A-Z]+ is deprecated and will be removed in Pillow 10:DeprecationWarning',  # keras
     "ignore:ml_dtypes.float8_e4m3b11 is deprecated.",  #FIXME: Can remove when jaxlib>=0.4.12
     "ignore:jsonschema.RefResolver is deprecated as of v4.18.0, in favor of the:DeprecationWarning",  # Issue #2139

--- a/src/pyhf/infer/calculators.py
+++ b/src/pyhf/infer/calculators.py
@@ -440,9 +440,15 @@ class AsymptoticCalculator:
                 qmu_A = tensorlib.power(self.sqrtqmuA_v, 2)
                 return (qmu - qmu_A) / (2 * self.sqrtqmuA_v)
 
-            # Use '<=' rather than '<' to avoid Issue #1992
+            # Use '<=' rather than '<' to avoid Issue #1992.
+            # sqrtqmuA_v == 0 when poi_test == asimov_mu (e.g. the poi=0 scan
+            # point); the q̃_mu > q̃_mu,A branch of arXiv:1007.1727 Eq. 16 is then
+            # 0/0, so route to the Gaussian-regime expression instead of
+            # dividing by zero (Issue #2722).
             teststat = tensorlib.conditional(
-                (sqrtqmu_v <= self.sqrtqmuA_v), _true_case, _false_case
+                (sqrtqmu_v <= self.sqrtqmuA_v) | (self.sqrtqmuA_v == 0),
+                _true_case,
+                _false_case,
             )
         return tensorlib.astensor(teststat)
 

--- a/src/pyhf/infer/calculators.py
+++ b/src/pyhf/infer/calculators.py
@@ -441,10 +441,13 @@ class AsymptoticCalculator:
                 return (qmu - qmu_A) / (2 * self.sqrtqmuA_v)
 
             # Use '<=' rather than '<' to avoid Issue #1992.
-            # sqrtqmuA_v == 0 when poi_test == asimov_mu (e.g. the poi=0 scan
-            # point); the q̃_mu > q̃_mu,A branch of arXiv:1007.1727 Eq. 16 is then
-            # 0/0, so route to the Gaussian-regime expression instead of
-            # dividing by zero (Issue #2722).
+            #
+            # sqrtqmuA_v == 0 happens when poi_test == asimov_mu
+            # (the poi=0 scan point) which results in the muhat < 0 branch
+            # of Equation 16 from https://arxiv.org/abs/1007.1727 (_false_case)
+            # being 0/0. To avoid this route this degenerate case to muhat >= 0
+            # (_true_case) (using bitwise OR) instead of dividing by zero
+            # (avoids Issue #2722).
             teststat = tensorlib.conditional(
                 (sqrtqmu_v <= self.sqrtqmuA_v) | (self.sqrtqmuA_v == 0),
                 _true_case,

--- a/src/pyhf/infer/calculators.py
+++ b/src/pyhf/infer/calculators.py
@@ -445,7 +445,7 @@ class AsymptoticCalculator:
             # sqrtqmuA_v == 0 happens when poi_test == asimov_mu
             # (the poi=0 scan point) which results in the muhat < 0 branch
             # of Equation 16 from https://arxiv.org/abs/1007.1727 (_false_case)
-            # being 0/0. To avoid this route this degenerate case to muhat >= 0
+            # being 0/0. To avoid this, route this degenerate case to muhat >= 0
             # (_true_case) (using bitwise OR) instead of dividing by zero
             # (avoids Issue #2722).
             teststat = tensorlib.conditional(

--- a/tests/test_infer.py
+++ b/tests/test_infer.py
@@ -683,6 +683,29 @@ def test_teststat_nan_guard():
     assert all(~np.isnan(result) for result in test_results)
 
 
+def test_teststatistic_poi0_no_divide_by_zero(hypotest_args):
+    # Issue #2722: at poi_test == asimov_mu (poi=0 for qtilde) sqrtqmuA_v == 0,
+    # so the q̃_mu > q̃_mu,A branch of arXiv:1007.1727 Eq. 16 is 0/0. The
+    # divide-by-zero only surfaces when the observed q̃_0 rounds to a tiny
+    # positive (e.g. macOS arm64); promote it to an error here independently of
+    # the project filterwarnings so the guard does not regress.
+    _, data, model = hypotest_args
+
+    calc = pyhf.infer.calculators.AsymptoticCalculator(data, model, test_stat="qtilde")
+    with warnings.catch_warnings():
+        warnings.filterwarnings("error", "divide by zero encountered")
+        teststat = calc.teststatistic(0.0)
+    assert np.isfinite(np.asarray(teststat))
+
+    # CLs at the background-only point must be finite and ~1 (mu=0 is never
+    # excluded), not NaN from a +inf test statistic.
+    with warnings.catch_warnings():
+        warnings.filterwarnings("error", "divide by zero encountered")
+        CLs_obs = pyhf.infer.hypotest(0.0, data, model, test_stat="qtilde")
+    assert np.isfinite(CLs_obs)
+    assert CLs_obs == pytest.approx(1.0, abs=1e-3)
+
+
 # TODO: Remove after pyhf v0.9.0 is released
 def test_deprecated_upperlimit(hypotest_args):
     with warnings.catch_warnings(record=True) as _warning:

--- a/tests/test_infer.py
+++ b/tests/test_infer.py
@@ -688,7 +688,7 @@ def test_teststatistic_poi0_no_divide_by_zero(hypotest_args):
     # At poi_test == asimov_mu (poi=0 for qtilde) sqrtqmuA_v == 0,
     # results in the muhat < 0 branch of Equation 16 from
     # https://arxiv.org/abs/1007.1727 being 0/0. This only occurs when the
-    # observed q̃_0 rounds to a tiny positive (happens on macOS amr64).
+    # observed q̃_0 rounds to a tiny positive (happens on macOS arm64).
     # Raise as an error to ensure this guard does not regress.
     _, data, model = hypotest_args
 

--- a/tests/test_infer.py
+++ b/tests/test_infer.py
@@ -684,11 +684,12 @@ def test_teststat_nan_guard():
 
 
 def test_teststatistic_poi0_no_divide_by_zero(hypotest_args):
-    # Issue #2722: at poi_test == asimov_mu (poi=0 for qtilde) sqrtqmuA_v == 0,
-    # so the q̃_mu > q̃_mu,A branch of arXiv:1007.1727 Eq. 16 is 0/0. The
-    # divide-by-zero only surfaces when the observed q̃_0 rounds to a tiny
-    # positive (e.g. macOS arm64); promote it to an error here independently of
-    # the project filterwarnings so the guard does not regress.
+    # Issue https://github.com/scikit-hep/pyhf/issues/2722
+    # At poi_test == asimov_mu (poi=0 for qtilde) sqrtqmuA_v == 0,
+    # results in the muhat < 0 branch of Equation 16 from
+    # https://arxiv.org/abs/1007.1727 being 0/0. This only occurs when the
+    # observed q̃_0 rounds to a tiny positive (happens on macOS amr64).
+    # Raise as an error to ensure this guard does not regress.
     _, data, model = hypotest_args
 
     calc = pyhf.infer.calculators.AsymptoticCalculator(data, model, test_stat="qtilde")

